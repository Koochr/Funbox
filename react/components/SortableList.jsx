import React from "react"
import PropTypes from "prop-types"
import { SortableContainer, SortableElement, arrayMove, } from "react-sortable-hoc"

const ListItem = SortableElement(({ value, sortIndex, renderData, }) =>
	<div className="sortable-list-item list-group-item list-group-item-light">
		{renderData(value, sortIndex)}
	</div>
)

const ItemsList = SortableContainer(({ items, renderData, }) => {
	return (
		<div className="list-group">
			{items.map((item, index) => (
				<ListItem
					key={`item-${index}`}
					index={index}
					sortIndex={index}
					value={item}
					renderData={renderData}
				/>
			))}
		</div>
	)
})

export default class WayPointsList extends React.Component {
  static propTypes = {
  	items: PropTypes.array.isRequired,
  	onOrderChange: PropTypes.func.isRequired,
  	renderData: PropTypes.func.isRequired,
  }

  onSortEnd = ({ oldIndex, newIndex, }) => {
  	this.props.onOrderChange(arrayMove(this.props.items, oldIndex, newIndex))
  }

  render() {
  	const { items, renderData, } = this.props
  	return <ItemsList items={items} onSortEnd={this.onSortEnd} renderData={renderData} />
  }
}
