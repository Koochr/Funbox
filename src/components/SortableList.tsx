import * as React from 'react';
import { SortableContainer, SortableElement, arrayMove } from 'react-sortable-hoc';
import { WayPoint, RenderListItem } from '../App';

interface SortableListProps {
  items: WayPoint[];
  onOrderChange: (wayPoints: WayPoint[]) => void;
  renderData: RenderListItem;
}

const ListItem = SortableElement(
  ({ value, sortIndex, renderData }: { value: WayPoint; sortIndex: number; renderData: RenderListItem }) => (
    <div className="sortable-list-item list-group-item list-group-item-light">{renderData(value, sortIndex)}</div>
  ),
);

const ItemsList = SortableContainer(
  ({ items, renderData }: { items: WayPoint[]; renderData: RenderListItem }) => (
    <div className="list-group">
      {items.map((item: WayPoint, index: number) => (
        <ListItem key={`item-${index}`} index={index} sortIndex={index} value={item} renderData={renderData} />
      ))}
    </div>
  ));

const SortableList: React.FunctionComponent<SortableListProps> = props => {
  const onSortEnd: ({ oldIndex, newIndex }: { oldIndex: number; newIndex: number }) => void =
        ({ oldIndex, newIndex }) => {
          props.onOrderChange(arrayMove(props.items, oldIndex, newIndex));
        };

  return <ItemsList items={props.items} onSortEnd={onSortEnd} renderData={props.renderData} />;
};

export default SortableList;
