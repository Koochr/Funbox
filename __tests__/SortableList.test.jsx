/* global it, expect */
import React from "react"
import { mount, shallow, configure, } from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import SortableList from "../react/components/SortableList.jsx"

configure({ adapter: new Adapter(), })

const items = ["item1", "item2", "item3",]

it("Should render correctly", () => {
	const component = shallow(
		<SortableList items={items} onOrderChange={() => {}} renderData={() => {}}/>
	)
	expect(component).toMatchSnapshot()
})

it("Should render needed components", () => {
	const component = mount(
		<SortableList items={items} onOrderChange={() => {}} renderData={(item) => <div>{item}</div>}/>
	)
	const list = component.find(".list-group")
	expect(list).toHaveLength(1)
	const listItems = list.find(".sortable-list-item")
	expect(listItems).toHaveLength(items.length)
	for (let i=0; i<items.length; i++) {
		expect(listItems.at(i).props().children.props.children).toStrictEqual(items[i])
	}
})
