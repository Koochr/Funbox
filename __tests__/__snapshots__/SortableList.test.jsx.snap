// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WayPointsList
    items={
      Array [
        "item1",
        "item2",
        "item3",
      ]
    }
    onOrderChange={[Function]}
    renderData={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "axis": "y",
      "distance": 0,
      "getHelperDimensions": [Function],
      "hideSortableGhost": true,
      "items": Array [
        "item1",
        "item2",
        "item3",
      ],
      "lockOffset": "50%",
      "lockToContainerEdges": false,
      "onSortEnd": [Function],
      "pressDelay": 0,
      "pressThreshold": 5,
      "renderData": [Function],
      "shouldCancelStart": [Function],
      "transitionDuration": 300,
      "useWindowAsScrollContainer": false,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "axis": "y",
        "distance": 0,
        "getHelperDimensions": [Function],
        "hideSortableGhost": true,
        "items": Array [
          "item1",
          "item2",
          "item3",
        ],
        "lockOffset": "50%",
        "lockToContainerEdges": false,
        "onSortEnd": [Function],
        "pressDelay": 0,
        "pressThreshold": 5,
        "renderData": [Function],
        "shouldCancelStart": [Function],
        "transitionDuration": 300,
        "useWindowAsScrollContainer": false,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
